src/pages/billing/_tests/CreateAzureBillingProjectModal.test.js

```diff
@@ -0,0 +1,238 @@
+  })
+
+  it.each([
+    { text: 'short', errors: [tooShortError] },
+    { text: '', errors: [tooShortError, badCharsError] },
+    { text: '**********', errors: [badCharsError] },
+    { text: 'aowejfawioefjaowiejfaoiwejfoaijw efoiawjefiajwefijaweoifjaweoijf', errors: [badCharsError, tooLongError] },
+    { text: 'ThisIs-a_suitableName', errors: [] }
+  ])('validates billing project name "$text"', ({ text, errors }) => {
+    const allErrors = [tooShortError, tooLongError, badCharsError, duplicateProjectError]
+
+    // Act - Insert billing project name
+    if (text === '') {
+      // Must first insert something and then replace with empty string.
+      fireEvent.change(getBillingProjectInput(), { target: { value: 'temp' } })
+    }
+    fireEvent.change(getBillingProjectInput(), { target: { value: text } })
+
+    // Assert - Verify expected error message(s) and presence of hint (hint will not show if there is an error).
+    if (_.isEmpty(errors)) {
+      expect(getBillingProjectHint()).not.toBeNull()
+      _.forEach(unexpectedError => expect(screen.queryByText(unexpectedError)).toBeNull(), allErrors)
+    } else {
+      expect(getBillingProjectHint()).toBeNull()
+      _.forEach(error => expect(screen.queryByText(error)).not.toBeNull(), errors)
+      _.forEach(unexpectedError => expect(screen.queryByText(unexpectedError)).toBeNull(), _.difference(allErrors, errors))
+    }
+    // Create always disabled because all fields must have valid inputs.
+    verifyDisabled(getCreateButton())
+  })
+
+  it('validates the subscription ID', () => {
+    // Arrange - Mock managed app Ajax call, should not be called
+    const listAzureManagedApplications = jest.fn(() => Promise.resolve())
+    Ajax.mockImplementation(() => { return { Billing: { listAzureManagedApplications } } })
+
+    // Assert - UUID error message should not initially be visible, even though subscription ID field is empty.
+    expect(screen.queryByText(invalidUuidError)).toBeNull()
+
+    // Act - Supply invalid UUID
+    fireEvent.change(getSubscriptionInput(), { target: { value: 'invalid UUID' } })
+
+    // Assert
+    expect(listAzureManagedApplications).not.toHaveBeenCalled()
+    expect(screen.queryByText(invalidUuidError)).not.toBeNull()
+    verifyDisabled(getCreateButton())
+    verifyDisabled(getManagedAppInput())
+  })
+
+  it('shows an error if there are no managed apps (valid subscription ID)', async () => {
+    // Arrange - Mock managed app Ajax call to return an empty list.
+    const subscriptionId = uuid()
+    const listAzureManagedApplications = jest.fn(() => Promise.resolve({ managedApps: [] }))
+    Ajax.mockImplementation(() => { return { Billing: { listAzureManagedApplications } } })
+
+    // Act - Supply valid UUID
+    await act(async () => {
+      await userEvent.type(getSubscriptionInput(), subscriptionId)
+    })
+
+    // Assert
+    expect(listAzureManagedApplications).toHaveBeenCalledWith(subscriptionId)
```

Comment:
I recommend pairing args assertions like this with a call count == 1 check.

Reply:
I'll keep that in mind…. I actually had the call count check first, but decided it was overkill. Note that if the mock isn't called, the error `toHaveBeenCalledWith` is very clear, so the only thing you are losing is checking that it isn't called more than once.